File: Sources/PUPSISPortalApp/Core/Markdown.swift
import Foundation

/// A minimal, Obsidian-flavored Markdown block model — enough for notes: ATX
/// headings (`#`…`######`), unordered bullets (`-`/`*`), task checkboxes
/// (`- [ ]` / `- [x]`), and plain paragraphs. Inline emphasis (`**bold**`,
/// `*italic*`, `` `code` ``) is left to `AttributedString(markdown:)` at render
/// time — this only splits the block structure SwiftUI's inline parser can't.
///
/// Pure and line-based on purpose, so it's testable without a view.
enum MarkdownBlock: Equatable {
    case heading(level: Int, text: String)
    case bullet(text: String)
    case checkbox(done: Bool, text: String)
    case paragraph(text: String)
    case blank
}

enum Markdown {
    static func parse(_ text: String) -> [MarkdownBlock] {
        text.components(separatedBy: "\n").map(block(from:))
    }

    private static func block(from line: String) -> MarkdownBlock {
        let trimmed = line.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return .blank }

        // Heading: 1–6 leading '#', then a space.
        if trimmed.first == "#" {
            let hashes = trimmed.prefix { $0 == "#" }.count
            let rest = trimmed.dropFirst(hashes)
            if hashes <= 6, rest.first == " " {
                return .heading(level: hashes, text: rest.trimmingCharacters(in: .whitespaces))
            }
        }

        // Bullet / checkbox: '- ' or '* '.
        for marker in ["- ", "* "] where trimmed.hasPrefix(marker) {
            let rest = String(trimmed.dropFirst(marker.count))
            if rest.hasPrefix("[ ] ") {
                return .checkbox(done: false, text: String(rest.dropFirst(4)))
            }
            if rest.lowercased().hasPrefix("[x] ") {
                return .checkbox(done: true, text: String(rest.dropFirst(4)))
            }
            return .bullet(text: rest)
        }

        return .paragraph(text: trimmed)
    }
}

File: windows/PUPSISPortal.Core/Markdown.cs
namespace PUPSISPortal.Core;

/// <summary>
/// A minimal, Obsidian-flavored Markdown block model — enough for notes: ATX
/// headings (`#`…`######`), unordered bullets (`-`/`*`), task checkboxes
/// (`- [ ]` / `- [x]`), and plain paragraphs. Inline emphasis (`**bold**`,
/// `*italic*`, `` `code` ``) is left to rendering at runtime — this only
/// splits the block structure that rendering can't handle.
///
/// Pure and line-based on purpose, so it's testable without a view.
/// </summary>
public class MarkdownBlock : IEquatable<MarkdownBlock>
{
    public enum Kind
    {
        Heading,
        Bullet,
        Checkbox,
        Paragraph,
        Blank,
    }

    public Kind Type { get; }
    public int Level { get; }
    public bool Done { get; }
    public string Text { get; }

    private MarkdownBlock(Kind type, int level = 0, bool done = false, string text = "")
    {
        Type = type;
        Level = level;
        Done = done;
        Text = text;
    }

    public static MarkdownBlock Heading(int level, string text) => new(Kind.Heading, level, false, text);
    public static MarkdownBlock Bullet(string text) => new(Kind.Bullet, 0, false, text);
    public static MarkdownBlock Checkbox(bool done, string text) => new(Kind.Checkbox, 0, done, text);
    public static MarkdownBlock Paragraph(string text) => new(Kind.Paragraph, 0, false, text);
    public static MarkdownBlock Blank => new(Kind.Blank);

    public override bool Equals(object? obj) => Equals(obj as MarkdownBlock);

    public bool Equals(MarkdownBlock? other)
    {
        if (other is null) return false;
        return Type == other.Type && Level == other.Level && Done == other.Done && Text == other.Text;
    }

    public override int GetHashCode() => HashCode.Combine(Type, Level, Done, Text);
}

/// <summary>
/// Parse Markdown text into blocks.
/// </summary>
public static class Markdown
{
    public static List<MarkdownBlock> Parse(string text)
    {
        return text.Split('\n')
            .Select(Block)
            .ToList();
    }

    private static MarkdownBlock Block(string line)
    {
        var trimmed = line.Trim();
        if (string.IsNullOrEmpty(trimmed))
            return MarkdownBlock.Blank;

        // Heading: 1–6 leading '#', then a space.
        if (trimmed[0] == '#')
        {
            var hashCount = 0;
            foreach (var c in trimmed)
            {
                if (c == '#') hashCount++;
                else break;
            }

            if (hashCount >= 1 && hashCount <= 6 && trimmed.Length > hashCount && trimmed[hashCount] == ' ')
            {
                return MarkdownBlock.Heading(hashCount, trimmed.Substring(hashCount + 1).Trim());
            }
        }

        // Bullet / checkbox: '- ' or '* '
        foreach (var marker in new[] { "- ", "* " })
        {
            if (trimmed.StartsWith(marker))
            {
                var rest = trimmed.Substring(marker.Length);
                if (rest.StartsWith("[ ] "))
                {
                    return MarkdownBlock.Checkbox(false, rest.Substring(4));
                }

                if (rest.StartsWith("[x] ", StringComparison.OrdinalIgnoreCase))
                {
                    return MarkdownBlock.Checkbox(true, rest.Substring(4));
                }

                return MarkdownBlock.Bullet(rest);
            }
        }

        return MarkdownBlock.Paragraph(trimmed);
    }
}
